netbox/dcim/api/serializers.py

```diff
@@ -31,6 +31,31 @@
 from .nested_serializers import *
 
 
+class BaseReadonlyDeviceMixin:
+
+    def get_fields(self, *args, **kwargs):
+        fields = super().get_fields(*args, **kwargs)
+
+        # get_extra_kwargs doesn't work if field explicitly declared on serializer...
+        if (self.instance):
+            fields[self.get_field_name()].read_only = True
+
+        return fields
+
+
+class ReadonlyDeviceMixin(BaseReadonlyDeviceMixin):
+
+    # can't be a field as mucks up serializer
+    def get_field_name(self):
+        return "device"
+
+
+class ReadonlyDeviceTypeMixin(BaseReadonlyDeviceMixin):
+
+    def get_field_name(self):
+        return "device_type"
+
+
```

Comment:
Can we simplify this a bit by doing something like this? Also, I'd really like to avoid introducing a new public method on the serializer as it could result in confusion.

```suggestion
class ReadonlyDeviceMixin:
    _field_name = 'device'

    def get_fields(self, *args, **kwargs):
        fields = super().get_fields(*args, **kwargs)

        if self.instance:
            fields[self._field_name].read_only = True

        return fields


class ReadonlyDeviceTypeMixin(ReadonlyDeviceMixin):
    _field_name = 'device_type'


```